framework/src/task_manager/task_manager_core.c

```diff
@@ -591,59 +605,173 @@ int taskmgr_get_handle_by_pid(int pid)
 	return TM_UNREGISTERED_APP;
 }
 
+static int taskmgr_check_broad_msg(int msg)
+{
+	if (tm_broadcast_msg[msg -1] == msg) {
+		return OK;
+	}
+	return TM_UNREGISTERED_MSG;
+}
+
+static tm_broadcast_t *taskmgr_search_broadcast_info(int msg, int handle)
+{
+	int ret;
+	sq_entry_t curr;
+
+	ret = taskmgr_check_broad_msg(msg);
+	if (ret == TM_UNREGISTERED_MSG) {
+		return NULL;
+	}
+	curr.flink = TM_BROADCAST_INFO_LIST(handle).head;
+	while (curr.flink != NULL) {
+		if (((tm_broadcast_t *)curr.flink)->msg == msg) {
+			return ((tm_broadcast_t *)curr.flink);
+		} else {
+			curr.flink = ((FAR sq_entry_t *)curr.flink)->flink;
+		}
+	}
+	return ((tm_broadcast_t *)curr.flink);
+}
+
 static void taskmgr_broadcast(int msg)
 {
 	int handle;
 	int fd;
 	int ret;
 	union sigval msg_broad;
+	tm_broadcast_t *broadcast_info;
 
 	fd = taskmgr_get_drvfd();
 	if (fd < 0) {
 		return;
 	}
-	msg_broad.sival_int = msg;
+
+	ret = taskmgr_check_broad_msg(msg);
+	if (ret == TM_UNREGISTERED_MSG) {
+		return;
+	}
+
 	for (handle = 0; handle < CONFIG_TASK_MANAGER_MAX_TASKS; handle++) {
 		if (TM_LIST_ADDR(handle) != NULL) {
 			ret = taskmgr_get_task_state(handle);
 			if (ret == TM_APP_STATE_STOP || ret == TM_APP_STATE_UNREGISTERED) {
 				continue;
 			}
-			ret = taskmgr_check_msg_mask(msg, handle);
-			if (ret != OK) {
+			broadcast_info = taskmgr_search_broadcast_info(msg, handle);
+			if (broadcast_info == NULL) {
 				continue;
 			}
 			ret = ioctl(fd, TMIOC_BROADCAST, TM_PID(handle));
 			if (ret != OK) {
 				continue;
 			}
+			msg_broad.sival_ptr = broadcast_info;
 			(void)sigqueue(TM_PID(handle), SIGTM_BROADCAST, msg_broad);
 		}
 	}
 }
 
+static void taskmgr_broadcast_msg_init(void)
+{
+	int chk_idx;
+	int msg_num;
+	int broadcast_max_num;
+
+#ifdef CONFIG_USER_DEFINED_BROADCAST
+	broadcast_max_num = TM_BROADCAST_USER_DEFINE_MAX;
+#else
+	broadcast_max_num = TM_BROADCAST_DEFINE_MAX;
+#endif
+	for (chk_idx = 0; chk_idx < broadcast_max_num + CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		msg_num = chk_idx + 1;
+		if (chk_idx < (broadcast_max_num - 1)) {
+			tm_broadcast_msg[chk_idx] = msg_num;
+		} else {
+			tm_broadcast_msg[chk_idx] = -1;
+		}
+	}
+}
+
 static int taskmgr_set_msg_cb(int type, void *data, int pid)
 {
 	int handle;
+	int ret;
+	int broadcast_max_num;
 
 	if (data == NULL) {
 		return TM_INVALID_PARAM;
 	}
-
 	handle = taskmgr_get_handle_by_pid(pid);
 	if (handle == TM_UNREGISTERED_APP) {
 		return handle;
 	}
-
 	if (type == TYPE_UNICAST) {
-		TM_UNICAST_CB(handle) = (_tm_unicast_t)data;
+		TM_UNICAST_CB(handle) = (_tm_cb_t)data;
 	} else {
-		TM_MSG_MASK(handle) = ((tm_broadcast_t *)data)->msg_mask;
-		TM_BROADCAST_CB(handle) = ((tm_broadcast_t *)data)->cb;
+#ifdef CONFIG_USER_DEFINED_BROADCAST
+	broadcast_max_num = TM_BROADCAST_USER_DEFINE_MAX;
+#else
+	broadcast_max_num = TM_BROADCAST_DEFINE_MAX;
+#endif
+		if (!((tm_broadcast_msg[0] == 1) || (tm_broadcast_msg[0] == -1))) {
```

Comment:
Positive is more easy to understand,
 ((tm_broadcast_msg[0] != 1) || (tm_broadcast_msg[0] != -1))

Reply:
When `taskmgr_broadcast_msg_init()` is called, tm_broadcast_msg[0] is set to 1 or -1.
Then, `((tm_broadcast_msg[0] != 1) || (tm_broadcast_msg[0] != -1))` is always true.
Therefore, `!(((tm_broadcast_msg[0] == 1) || (tm_broadcast_msg[0] == -1)))` can not be changed into `((tm_broadcast_msg[0] != 1) || (tm_broadcast_msg[0] != -1))`